### DISCUSSION TITLE
ESPNow coding issue between ESP8266 & ESP32

### DISCUSSION THREAD

**Discussion by mmliam** [micropython/micropython#9128] [Using MicroPython]
ESPNow code runs in ESP8266, throws an error in ESP32.

Code:
```
eNow = espnow.ESPNow()
enRcvr = b'\x01\x02\x03\x04\x05\x06'
devState = bytes('On', 'ascii')
ackRcvd = eNow.send(enRcvr,devState,True)

```
The full code runs without error in an ESP8266

In an ESP32 it throws this error:

> Traceback (most recent call last):
  File "main.py", line 29, in <module>
OSError: (-12393, 'ESP_ERR_ESPNOW_NOT_FOUND')
MicroPython v1.19.1-espnow-6-g44f65965b on 2022-07-09; ESP32 module with ESP32

 Line 29 is the eNow.send function.
Both the ESP8266 & ESP32 are using their respective v1.19.1-espnow builds.

Anyone know what's causing the error for the ESP32?

Mike

**Comment by davefes** [comment]
`import espnow`

works for me on ESP32 Generic and SPIRAM

**Reply by mmliam** [reply]
I haven't shown all the code details, but I have "import espnow". Remember, I said the code works for the ESP8266 that has the same build; a missing import would have been flagged.

I have a feeling the error is more subtle.

**Comment by davefes** [comment]
I wasn't going to assume that they both the same.

Where are you making the interface active before trying to send?

```
e0 = espnow.ESPNow()
e0.active(True)
e0.add_peer(config.repeater)
e0.send()
```
N.B. e0.init() has been replaced with e0.active(True)

**Reply by mmliam** [reply]
Basically the same order.

**Comment by davefes** [comment]
I still don't see where you are making it active.

**Comment by mmliam** [comment]
As I said I've only posted the reliant code-snippet; and given that the full code runs without error in an ESP8266 all the basic coding; imports, and orders must not be the problem.

**Comment by davefes** [comment]
e0.init() might behave differently between the ESP8266 and ESP32 v1.19.1 images.  Where ever you are making the interface active in the ESP8266 code try using e0.active(True) instead.

**Reply by mmliam** [reply]
Actually, according to Glenn, the .init also performs an .active.

I have ESPNow running on an ESP32, but only as a receiver. I really don't expect to be using an ESP32 as a sender, but I was going to run  simple test when this unexpected error popped up.

**Reply by davefes** [reply]
Yes.

You have registered the peer you are trying to send to?

**Comment by davefes** [comment]
I have found:
`    print (e0.get_peers())`
helpful

**Reply by mmliam** [reply]
BTW: Just for kicks, I tried e.active(True). As expected it didn't make any difference; there's definitely something weird going on.

**Comment by davefes** [comment]
For both ends of the link you are running power_saving(0) (the default) and both are STA?

**Reply by mmliam** [reply]
Don't see have that would result in an interpreter error. The unit in question is actively monitored in Thonny.

**Comment by davefes** [comment]
Sent you a PM on the old forum

**Comment by glenn20** [comment]
@mmliam I know we resolved this outside this discussion - but I thought I'd reply here as well for the benefit of others who may encounter this issue.

> In an ESP32 it throws this error:
> 
> > Traceback (most recent call last):
> > File "main.py", line 29, in 
> > OSError: (-12393, 'ESP_ERR_ESPNOW_NOT_FOUND')

This error is returned by the Espressif software stack when sending to a device which has not been registered with `ESPNow.add_peer(mac_address)`. (See [Micropython ESPNow docs](https://micropython-glenn20.readthedocs.io/en/latest/library/espnow.html#exceptions) and [Espressif ESP-Now documentation](https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/network/esp_now.html#_CPPv412esp_now_sendPK7uint8_tPK7uint8_t6size_t))

The Espressif ESP8266 SDK does not enforce this requirement, but the ESP32 IDF does.

While the Espressif ESP-Now API is very similar on the esp8266 and the esp32, they are not identical and even where they are similar, there can be subtle differences in behaviour. If you want your code to be portable between esp8266 and esp32 you should always register the peer with add_peer() before sending messages to that peer.
